fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,560 @@
+        }
+        ++tokenIt; // consume to be COMMA, the node value, or the RPAREN_SQ
+    }
+    return currentValue;
+}
+
+/* Parse and build the list of the node argument
+ * 
+ * @param tokenIt The tokens contains the node argument info
+ * @return The node argument linked list
+ */
+ASTNodeArg* parseArgList(TokenIter &tokenIt) {
+    ASTNodeArg* currentArg = NULL;
+    ASTNodeArg* argList = NULL;
+
+    while (tokenIt->getType() != Token::LPAREN && tokenIt->getType() != Token::RPAREN) {
+        if (tokenIt->getType() == Token::ID && tokenIt.peek().getType() == Token::EQUALS) {
+            char * node_args_name = new char[tokenIt->getValue().size() + 1];
+            strcpy(node_args_name, tokenIt->getValue().c_str());
+            ++tokenIt; // consume EQUALS
+            ++tokenIt; // consume to be LPAREN_SQ(the start of value list) or single node value
+            if (tokenIt->getType() == Token::LPAREN_SQ) {
+                ++tokenIt;
+                currentArg = createNodeArg(node_args_name, parseValueList(tokenIt), NULL);
+            } else {
+                try {
+                    currentArg = createNodeArg(node_args_name, buildNodeValue(*tokenIt), NULL);
+                } catch (ParserFailure &f) {
+                    std::cerr << f.what() << std::endl;
+                }
+            }
+        } else if (checkValue(tokenIt)) {
+            try {
+                currentArg = createNodeArg("", buildNodeValue(*tokenIt), NULL);
+            } catch (ParserFailure &f) {
+                std::cerr << f.what() << std::endl;
+            }
+        } else {
+            throw ParserFailure("Incorrect type for argument, current value is " + tokenIt->getValue());
+        }
+        if (argList == NULL) {
+            argList = currentArg;
+        } else {
+            appendSiblingArg(argList, currentArg);
+        }
+        ++tokenIt; // consume to be the start of the next Arg (could be single node value or ID followed by BECOMES)
+    }
+    return argList;
+}
+
+/* Parse and build the AST
+ * 
+ * @param tokenIt The tokens contains the AST info
+ * @return The root of the AST
+ */
+ASTNode* buildAST(TokenIter &tokenIt) {
+    ASTNode* siblingsList = NULL;
+    while (!tokenIt.isEnd() && tokenIt->getType() == Token::LPAREN) {
+        ASTNodeArg* argList = NULL;
+        ASTNode* currentchild = NULL;
+        ASTNode* childrenList = NULL;
+        ++tokenIt; // consume to be ID
+        if (tokenIt.isEnd()) {
+            throw ParserFailure("Token stream ended unexpectedly after LPAREN");
+        } else if (tokenIt->getType() != Token::ID) {
+            throw ParserFailure("The next token of LPAREN should be IDENTIFIER, while current next value is " +  tokenIt.peek().getValue());
+        }
+        char* node_name = new char[tokenIt->getValue().size() + 1];
+        strcpy(node_name, tokenIt->getValue().c_str());
+        ++tokenIt; // consume to be the start of argList/nodeList or RPAREN
```

Comment:
Again, I think it makes more sense to specify which token was "removed from the input stream" instead of when thing we are expecting next.